lib/ddtrace/sampler.rb

```diff
@@ -107,6 +108,8 @@ def update(rate_by_service)
           @sampler[key] ||= RateSampler.new(rate)
           @sampler[key].sample_rate = rate
         end
+
+        Diagnostics::Health.metrics.sampling_service_cache_length(rate_by_service.length)
```

Comment:
The number of `service,env` keys in the response from the agent?

What does this tell us?

Reply:
It tells us if our cache is getting massive, because it creates a sampler per service. Last I checked, the agent sends back *all services its ever received*, not just the ones from this application, so if you have an agent that serves 1000 services, it will stuff Ruby full of 999 samplers that it doesn't necessarily need.

Reply:
I think it only returns back the ones you sent in the payload.

Reply:
No, that's not the case, not when I last checked. I've seen it in outputs from users.

Reply:
Hmm, it is all, I could have sworn I had seen it return only a subset before.

https://github.com/DataDog/datadog-agent/blob/780caa2855a237fa731b78a1bb3ead5492f0e5c6/pkg/trace/api/responses.go#L73

Still not sure if this is super useful. Not sure it tells us anything interesting, even if there were 2k services, the overhead of having that many services shouldn't be bad and is information we can likely get from the agent anyways.

Reply:
It might be bad; each entry defines another sampler so it has the potential to grow larger faster. Also, as something that remains permanently in global memory, it's probably a good idea to measure things like this, then if they don't prove useful get rid of them.

I know this has caused a problem in at least one user app before (when they had a ton of services being created), so it's plausible it might help identify an issue. (It's already helped us identify a bug with priority sampling, so it's off to a productive start.)